No way to fetch values of named Tags in IMDS Instance Metadata

Hi!

Just got to trying the 2.0.0 release, well done :muscle: Sincerely appreciated.

An immediate issue with EC2 Metadata: using amazonka, it's **impossible to fetch actual values** of instance `Tags`.

To showcase what I mean, here I'm primarily interested in the `Name` instance tag:
```
$ curl http://instance-data/latest/meta-data/tags/instance/Name
app-dev-04
# ↑↑↑ the Name tag value I need to fetch in app

$ curl http://instance-data/latest/meta-data/tags/instance/
Cost_center
Customer
Environment
Name
Type
# ↑↑↑ all instance tags, \n-separated... keys only! no values!
```

Given this:

https://github.com/brendanhay/amazonka/blob/e081929d5a81f30397c9c9a219b154da1a922fb8/lib/amazonka/src/Amazonka/EC2/Metadata.hs#L576-L587

— it's only possible to do the 2nd curl, obtaining the result `"Cost_center\nCustomer\nEnvironment\nName\nType"` — which for my purpose is (obviously, I hope) utterly useless. The 1st curl which I need, is impossible to express.

I've attempted to work around by:
```haskell
data InstanceNameTag = InstanceNameTag
instance AmazonkaCore.ToText InstanceNameTag where toText _ = "instance/Name"

    {- ... -}
    instanceName <- IMDS.metadata man InstanceNameTag
```
— but ofcourse, it's a type error :confused: 
```
src/Application/Metrics.hs:159:37: error:
    • Couldn't match expected type ‘Metadata’
                  with actual type ‘InstanceNameTag’
    • In the second argument of ‘metadata’, namely ‘InstanceNameTag’
      In a stmt of a 'do' block:
        instanceName <- metadata man InstanceNameTag
      In the expression:
        do instanceId <- metadata man InstanceId <&> decodeUtf8
           machineName <- metadata man (Tags Instance) <&> showT
           instanceName <- metadata man InstanceNameTag
           let (internalIpv4, internalPort) = ...
           ....
    |       
159 |   instanceName <- IMDS.metadata man InstanceNameTag
    |                                     ^^^^^^^^^^^^^^^
```

--------------------------------------------------------------------------

Would it sound sensible to add a constructor like so
```haskell
data Tags
  = Instance
  | InstanceTag !Text
```
— to allow querying any named tag using amazonka? Or extend `Metadata` instead?

Or am I missing some other way to extend here on the side of library-user? 

I can develop a patch and file a PR if any of above sounds any good. Just let me know.

<!-- an additional issue is:
    Yes of course, this has to run after
      aws ec2 modify-instance-metadata-options --instance-metadata-tags enabled --instance-id i-0123456789abcdef0
    — however it's supposed to not be requried for IMDSv2, implemented in #831.
-->

--- a/lib/amazonka/CHANGELOG.md
+++ b/lib/amazonka/CHANGELOG.md
@@ -14,6 +14,7 @@
   Before this fix you had to create an empty credentials file for these methods to work correctly.
 - `amazonka`: During logging, do not print bytestrings that have unprintable characters
 [\#952](https://github.com/brendanhay/amazonka/pull/952)
+- `amazonka`: Support reading values of IMDS metadata tags, add `Metadata.Other` constructor [\#955](https://github.com/brendanhay/amazonka/pull/955)
 
 
 ## [2.0.0](https://github.com/brendanhay/amazonka/tree/2.0.0)

--- a/lib/amazonka/CHANGELOG.md
+++ b/lib/amazonka/CHANGELOG.md
@@ -14,7 +14,8 @@
   Before this fix you had to create an empty credentials file for these methods to work correctly.
 - `amazonka`: During logging, do not print bytestrings that have unprintable characters
 [\#952](https://github.com/brendanhay/amazonka/pull/952)
-- `amazonka`: Support reading values of IMDS metadata tags, add `Metadata.Other` constructor [\#955](https://github.com/brendanhay/amazonka/pull/955)
+- `amazonka`: Support reading values of IMDS metadata tags, add `Metadata.Other` constructor
+[\#955](https://github.com/brendanhay/amazonka/pull/955)
 
 
 ## [2.0.0](https://github.com/brendanhay/amazonka/tree/2.0.0)

--- a/lib/amazonka/src/Amazonka/EC2/Metadata.hs
+++ b/lib/amazonka/src/Amazonka/EC2/Metadata.hs
@@ -217,9 +217,9 @@ data Metadata
   | -- | Any other piece of metadata specified by arbitrary key.
     --
     -- This is provided for forward compatibility: should there ever appear any gap
-    -- between a future version of AWS API and what (possibly an outdated version of)
-    -- this library covers -- the constructor gives the option to work around.
-    XtendedMeta !Text
+    -- between a future version of IMDS API and what this library covers,
+    -- this constructor provides a workaround.
+    Other !Text
   deriving stock (Eq, Ord, Show, Generic)
 
 instance ToText Metadata where
@@ -258,7 +258,7 @@ instance ToText Metadata where
       Services m -> "services/" <> toText m
       Spot m -> "spot/" <> toText m
       Tags m -> "tags/" <> toText m
-      XtendedMeta m -> m
+      Other m -> m
 
 -- | Metadata keys for @autoscaling/*@.
 data Autoscaling

--- a/lib/amazonka/src/Amazonka/EC2/Metadata.hs
+++ b/lib/amazonka/src/Amazonka/EC2/Metadata.hs
@@ -580,18 +580,24 @@ instance ToText Spot where
     SInstanceAction -> "instance-action"
     STerminationTime -> "termination-time"
 
--- | Metadata keys for @tags/*@.
+-- | Instance Metadata tags: @tags/instance@, @tags/instance/*@.
+--
+-- See [these 2 curl examples](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html#instance-metadata-ex-7)
+-- in AWSEC2 User Guide.
+--
+-- Only available if you [explicitly allow access](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html#allow-access-to-tags-in-IMDS)
+-- to instance metadata tags.
 data Tags
-  = -- | The instance tags associated with the instance. Only
-    -- available if you explicitly allow access to tags in instance
-    -- metadata. For more information, see
-    -- [Allow access to tags in instance metadata](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/Using_Tags.html#allow-access-to-tags-in-IMDS).
+  = -- | The keys of available tags, @\n@-separated, if enabled in instance settings.
     Instance
+    -- | A specific named tag in Instance Metadata, if enabled in instance settings.
+  | InstanceTag !Text
   deriving stock (Eq, Ord, Show, Generic)
 
 instance ToText Tags where
   toText = \case
     Instance -> "instance"
+    InstanceTag t -> "instance/" <> t
 
 latest :: Text
 latest = "http://169.254.169.254/latest/"

--- a/lib/amazonka/src/Amazonka/EC2/Metadata.hs
+++ b/lib/amazonka/src/Amazonka/EC2/Metadata.hs
@@ -214,6 +214,12 @@ data Metadata
     Spot !Spot
   | -- | See: 'Tags'
     Tags !Tags
+  | -- | Any other piece of metadata specified by arbitrary key.
+    --
+    -- This is provided for forward compatibility: should there ever appear any gap
+    -- between a future version of AWS API and what (possibly an outdated version of)
+    -- this library covers -- the constructor gives the option to work around.
+    XtendedMeta !Text
   deriving stock (Eq, Ord, Show, Generic)
 
 instance ToText Metadata where
@@ -252,6 +258,7 @@ instance ToText Metadata where
       Services m -> "services/" <> toText m
       Spot m -> "spot/" <> toText m
       Tags m -> "tags/" <> toText m
+      XtendedMeta m -> m
 
 -- | Metadata keys for @autoscaling/*@.
 data Autoscaling

